### ISSUE TITLE
ERROR: Could not build wheels for ConfigSpace which use PEP 517 and cannot be installed directly

### ISSUE THREAD

**Issue by draconcio12** [automl/ConfigSpace#180]
Using python 3.7 can't install it.

**Comment by mfeurer** [comment]
Could you please give more details:

* the error message and traceback?
* python version and operating system
* cython version
* compiler

**Comment by draconcio12** [comment]
I try with Ubuntu and Elementary (same error) 
python 3.8
gcc 7.5.0
pip 21.0.1
cython 0.29.22


ERROR: Command errored out with exit status 1:
   command: /home/hector/Proyectos/fx/bin/python3.7 /home/hector/Proyectos/fx/lib/python3.7/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /tmp/tmp27872umn
       cwd: /tmp/pip-install-zpgb6amw/configspace_67c28ed5c85947cc9578862a481c0816
  Complete output (48 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.7
  creating build/lib.linux-x86_64-3.7/ConfigSpace
  copying ConfigSpace/__init__.py -> build/lib.linux-x86_64-3.7/ConfigSpace
  copying ConfigSpace/__version__.py -> build/lib.linux-x86_64-3.7/ConfigSpace
  copying ConfigSpace/exceptions.py -> build/lib.linux-x86_64-3.7/ConfigSpace
  creating build/lib.linux-x86_64-3.7/ConfigSpace/read_and_write
  copying ConfigSpace/read_and_write/pcs_new.py -> build/lib.linux-x86_64-3.7/ConfigSpace/read_and_write
  copying ConfigSpace/read_and_write/irace.py -> build/lib.linux-x86_64-3.7/ConfigSpace/read_and_write
  copying ConfigSpace/read_and_write/pcs.py -> build/lib.linux-x86_64-3.7/ConfigSpace/read_and_write
  copying ConfigSpace/read_and_write/__init__.py -> build/lib.linux-x86_64-3.7/ConfigSpace/read_and_write
  copying ConfigSpace/read_and_write/json.py -> build/lib.linux-x86_64-3.7/ConfigSpace/read_and_write
  creating build/lib.linux-x86_64-3.7/ConfigSpace/nx
  copying ConfigSpace/nx/release.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx
  copying ConfigSpace/nx/exception.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx
  copying ConfigSpace/nx/__init__.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx
  creating build/lib.linux-x86_64-3.7/ConfigSpace/nx/classes
  copying ConfigSpace/nx/classes/digraph.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx/classes
  copying ConfigSpace/nx/classes/__init__.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx/classes
  copying ConfigSpace/nx/classes/graph.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx/classes
  creating build/lib.linux-x86_64-3.7/ConfigSpace/nx/algorithms
  copying ConfigSpace/nx/algorithms/cycles.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx/algorithms
  copying ConfigSpace/nx/algorithms/dag.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx/algorithms
  copying ConfigSpace/nx/algorithms/__init__.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx/algorithms
  creating build/lib.linux-x86_64-3.7/ConfigSpace/nx/algorithms/components
  copying ConfigSpace/nx/algorithms/components/__init__.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx/algorithms/components
  copying ConfigSpace/nx/algorithms/components/strongly_connected.py -> build/lib.linux-x86_64-3.7/ConfigSpace/nx/algorithms/components
  running build_ext
  cythoning ConfigSpace/hyperparameters.pyx to ConfigSpace/hyperparameters.c
  cythoning ConfigSpace/forbidden.pyx to ConfigSpace/forbidden.c
  cythoning ConfigSpace/conditions.pyx to ConfigSpace/conditions.c
  cythoning ConfigSpace/c_util.pyx to ConfigSpace/c_util.c
  skipping 'ConfigSpace/util.c' Cython extension (up-to-date)
  skipping 'ConfigSpace/configuration_space.c' Cython extension (up-to-date)
  building 'ConfigSpace.hyperparameters' extension
  creating build/temp.linux-x86_64-3.7
  creating build/temp.linux-x86_64-3.7/ConfigSpace
  x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fdebug-prefix-map=/build/python3.7-a56wZI/python3.7-3.7.10=. -fstack-protector-strong -Wformat -Werror=format-security -g -fdebug-prefix-map=/build/python3.7-a56wZI/python3.7-3.7.10=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.7m -I/home/hector/Proyectos/fx/include/python3.7m -I/tmp/pip-build-env-0i5g5lcd/overlay/lib/python3.7/site-packages/numpy/core/include -c ConfigSpace/hyperparameters.c -o build/temp.linux-x86_64-3.7/ConfigSpace/hyperparameters.o
  ConfigSpace/hyperparameters.c:6:10: fatal error: Python.h: No such file or directory
   #include "Python.h"
            ^~~~~~~~~~
  compilation terminated.
  /tmp/pip-build-env-0i5g5lcd/overlay/lib/python3.7/site-packages/setuptools/dist.py:645: UserWarning: Usage of dash-separated 'description-file' will not be supported in future versions. Please use the underscore name 'description_file' instead
    % (opt, underscore_opt))
  error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for ConfigSpace
Failed to build ConfigSpace
ERROR: Could not build wheels for ConfigSpace which use PEP 517 and cannot be installed directly

**Comment by mfeurer** [comment]
Alright, it looks like you're running a system installation of Python without a virtual environment or conda environment. The error is that the file `Python.h` is missing, which you can get on Ubuntu by doing `apt install python3-dev`. 

I just added the missing package to the installation instructions: https://github.com/automl/ConfigSpace/blob/master/docs/source/index.rst#installation

Could you please let me know if that works?

**Comment by draconcio12** [comment]
Still not working, same error.

**Comment by kashankrm** [comment]
Hi, I had a similar issue. I was using Windows Subsystem Linux. Once i installed WSL2 this error was solved. If you are using WSL try upgrading or using a dual boot linux otherwise please ignore this.

**Comment by kashankrm** [comment]
Hey I figured out why you might be having this issue, so the fix is still to install python3-dev but you need to install the correct version so if you have say python3(in ubuntu 20 LTS this is same as python3.8) and python3.7 and if you created virtualenv with python3.7 then you need to install python3.7-dev. 

I ran into same issue and that was the fix. 
TLDR: if you are using system provided python then do `apt install python3-dev` otherwise if you are using a different version, install for that version (eg `apt install python3.7-dev`)

**Comment by mfeurer** [comment]
@draconcio12 does the answer from @kashankrm work for you?

**Comment by draconcio12** [comment]
Hello @mfeurer  I try  @kashankrm but doesn't work for me, so I decided to use an miniconda enviroment and works well.
Thanks for the answeres, I really appreciate it

**Comment by mfeurer** [comment]
Thanks for the update. I'll close the issue as you can use conda by now.
